src/Microsoft.AspNetCore.Blazor/Components/TreeParameterState.cs

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Blazor.Rendering;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Blazor.Components
+{
+    internal readonly struct TreeParameterState
+    {
+        private readonly static IDictionary<Type, ReflectedTreeParameterInfo[]> _cachedInfos
+            = new ConcurrentDictionary<Type, ReflectedTreeParameterInfo[]>();
+
+        public string LocalName { get; }
+        public ProviderBase FromProvider { get; }
+
+        public TreeParameterState(string localName, ProviderBase fromProvider)
+        {
+            LocalName = localName;
+            FromProvider = fromProvider;
+        }
+
+        public static ICollection<TreeParameterState> FindTreeParameters(ComponentState componentState)
```

Comment:
nit: consider `IReadOnlyList<>`  or just a List or array if this is an implementation detail. Those types are way more efficient to consume.

Reply:
It was changed to `IReadOnlyList<>` in a subsequent commit.